Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemController.java

```diff
@@ -77,16 +77,22 @@
     public List<?> getProblems(@RequestHeader(value = "Authorization", required = false) String authorization,
                                @RequestParam(value = "tags", required = false) String[] tags,
                                @RequestParam(value = "page", defaultValue = "0") int page,
+                               @RequestParam(value = "visible", defaultValue = "true") boolean visible,
+                               @RequestParam(value = "invisible", defaultValue = "true") boolean invisible,
+                               @RequestParam(value = "archive", required = false) Boolean archive,
                                @RequestParam(required = false) int[] ids) {
         var token = tokenService.parseBearerTokenAndValidate(authorization);
+        // only admin can view the invisible problems
         boolean includeInvisibleProblems = token.isAdmin();
         if (ids != null) {
             var presenter = new GetProblemsPresenter();
             getProblemsUseCase.execute(new GetProblemsUseCase.Request(includeInvisibleProblems, ids), presenter);
             return presenter.present();
         } else {
+            archive = archive == null ? null : archive && token.isAdmin();
+            invisible = invisible && includeInvisibleProblems;
```

Comment:
Since only the admin can query archived and invisible problems, we can refine the code as follows (includes the comments). The following code has more readability.
```java
// only admin can view the archived problems
archive = archive == null ? null : archive && token.isAdmin();

// only admin can view the invisible problems
invisible = invisible && token.isAdmin();
```